Interface Changes

alex-connolly opened the pull request.

alex-connolly committed:
First pass interface changes, full restructure. Untested, likely still buggy.

suluvai reviewed (commented):
*Interface*
- I like the concept of single package at top-level `immutable` although would like if named `imx` so that the import path will look like `github.com/immutable/imx-core-sdk-golang/imx`  
- All the root level `.go` files would be nice to be moved into `imx` folder. It will look clean with out the pollution from other files at root level.
- Would like to still keep the corresponding workflows to their respective files while still being single package.
- Love the organisation of signers, being new to company/concept of signers at the beginning we did not have the same understanding. Makes sense now.
- We do not have interface methods defined in `interface.go` would be good to have interface defined and then implement the interface.

*Generated Code*
- Just a question, are we still trying to limit the public API functions from autogenerated code or exporting all?
- Having the generated code under generated folder provided that distinction it is not to be edited. We lose that clarity.

suluvai commented on interface.go:
```diff
@@ -0,0 +1,97 @@
+	api.TokensApi
+	api.CollectionsApi
+	api.BalancesApi
+}
```
This will make interface implemented by members accessible by `Client` object. If there is a name clash for what ever reason this can break. 

When they are added as member vars, I prefer the clarity it provides in terms of the functions that are available for a particular endpoint.

alex-connolly replied:
I've gone with how we structured it in the TS SDK - there are definitely tradeoffs, but we should be consistent. On the flattening, I actually prefer flatter - otherwise people have to already understand the schema (e.g. balances vs. tokens vs. assets)

suluvai replied:
I feel the flat structure might confuse more. It will be more specific. if a user is only focusing on the relevant functions to get a particular job done like deposit or minting or assets, instead of understanding all the functions in flatter structure. No ambiguity when they are grouped accordingly.

suluvai replied:
Yup, will try to maintain consistency. For limiting the public interface functions we might not want to expose them at top level. We only want to expose the limited interface.

suluvai commented on contracts/IERC165.sol:
```diff
@@ -1,25 +0,0 @@
-     * This function call must use less than 30 000 gas.
-     */
-    function supportsInterface(bytes4 interfaceId) external view returns (bool);
-}
```
Sorry, I have checked this since you asked. 

This is required as ERC721 inherits from ERC165 token.

alex-connolly replied:
Not required - we only need to include functions we actually want to use, won't break anything

suluvai commented on ethereumutil/gas_params.go:
```diff
@@ -1,41 +0,0 @@
-		GasMultiplier: DefaultGasMultiplier,
-		MaxGasPrice:   big.NewInt(int64(DefaultMaxGasPrice)),
-	}
-}
```
I do not see this code moved to anywhere else? 

Does that mean we do not need to provide the user an option to configure these params? In that case does it use default values for everyone?

alex-connolly replied:
I think we need to provide a method to configure these params in the workflows where we initiate transactions, I'll add a commit

alex-connolly replied:
Added, let me know what you think!

zatlite replied:
This is perfect!

zatlite committed:
fix: DecodeBig expects 0x prefix

zatlite committed:
feat: NewClient sets API server according to Environment

Users will not be able to and will not have to pass in ContextServerIndex from context.

zatlite committed:
lint: update removed directories

zatlite committed:
lint: minor fixes

zatlite committed:
lint: remove unused return value

zatlite committed:
feat: implement CreateBatchNftTransfer

namespace clash on Client.CreateTransfer resolved by specifying Client.TransferApi.CreateTransfer

alex-connolly committed:
quick update

alex-connolly committed:
Adding transaction overrides to transaction workflows

zatlite committed:
fix(examples): pass in override params

- Deposit and CompleteWithdrawal interfaces updated to include `overrides *bind.TransactOpts`

alex-connolly committed:
updating stark package to match interface in typescript sdk

alex-connolly committed:
remove unused package

zatlite committed:
lint: remove empty line

`go mod tidy` in examples/workflows dir

zatlite committed:
chore: remove server selection by context from examples

this is not possible anymore with NewClient which sets a single server
on the api.Configuration

suluvai committed:
Moved root level go files to imx package.

suluvai committed:
Moved api to be under imx and moved contracts to be internal only

suluvai committed:
Fixed import paths for docs

suluvai committed:
Split each workflow into its own file

suluvai committed:
Updated linter paths

zatlite commented on imx/examples/workflows/main.go:
```diff
@@ -0,0 +1,103 @@
+
+	cfg := imx.Config{
+		APIConfig:   configuration,
+		EthereumRPC: "https://eth-ropsten.alchemyapi.io/v2/" + alchemyAPIKey,
```
Should this url be part of config Environment constants?

suluvai replied:
We will need to pass in the alchemyApiKey value, that is how it was previously until the changes in this PR.

suluvai replied:
Addressed

zatlite commented on imx/examples/workflows/prepareWithdrawal.go:
```diff
@@ -1,106 +1,99 @@
-	withdrawalsWorkflow "github.com/immutable/imx-core-sdk-golang/workflows/withdrawals"
+	"github.com/immutable/imx-core-sdk-golang/imx"
+	"github.com/immutable/imx-core-sdk-golang/imx/api"
+	"github.com/immutable/imx-core-sdk-golang/imx/internal/convert"
```
I think it would be better to remove the internal imports as SDK users will not be able to import from internal. We don't need the denomination conversion to demonstrate the example.

suluvai replied:
But it provides the clarity of what a user need to do, otherwise a user will have to go through the same journey as we did when we were testing this.

zatlite replied:
I think both issues raised here are worth resolving.

suluvai replied:
This will be addressed separately in subsequent PRs

zatlite commented on imx/signers/stark/signer.go:
```diff
@@ -1,69 +1,79 @@
-// Package stark implements utilities for generating signers.L2Signer
+// Package stark implements utilities for generating immutable.L2Signer
```
```suggestion
// Package stark implements utilities for generating imx.L2Signer
```

suluvai committed:
Moved examples to be under imx and fixes to compile code

suluvai committed:
shell script to generate contracts go code

suluvai committed:
Read me updates

suluvai committed:
Fixed readme and golang ci linter paths

suluvai commented on imx/burn.go:
```diff
@@ -18,32 +16,31 @@ type GetSignableBurnRequest struct {
 // We have added burn functions in the SDK for clarity, and these are just wrappers for standard transfer functions to BurnAddress.
 // Using the transfer functions with the burn address as a recipient will also work just the same.
-func Burn(
+func (c *Client) Burn(
```
ok, do you know what was the reason?

zatlite replied:
probably because burn is just transfer.

Kastet replied:
correct

suluvai committed:
PR feedback to move the EthereumRPC url to be part of environment

zatlite commented on README.md:
```diff
@@ -61,30 +61,45 @@ func main() {
+)
+
+func main() {
+   // L1 credentials, supply your signerPrivateKey string.
```
please address indenting issues in code samples here and more below.

zatlite commented on imx/examples/main.go:
```diff
@@ -0,0 +1,105 @@
+	}
+	defer c.EthClient.Close()
+
+	chainID, err := c.EthClient.ChainID(ctx)
```
This is not necessary now because you can get chainID from `Environment`.

zatlite commented on imx/interface.go:
```diff
@@ -0,0 +1,101 @@
+	api.BalancesApi
+}
+
+func NewClient(cfg Config) (*Client, error) {
```
Please add godoc comments to this file and address the lint issue.

suluvai committed:
PR feedback

zatlite commented on README.md:
```diff
@@ -107,7 +122,7 @@ The Core SDK includes classes that interact with the Immutable X APIs.
     "fmt"
-    "github.com/immutable/imx-core-sdk-golang/generated/api"
+    "github.com/immutable/imx-core-sdk-golang/imx/api"
     "github.com/immutable/imx-core-sdk-golang/config"
```
This import path and a few others like workflow no longer exist.
```
    "github.com/immutable/imx-core-sdk-golang/config"
```
Also, all the code samples need to update to the new way of calling everything from `imx` and accessing the methods from `Client` instead.

The samples should not use the `api` directly and should instead be calling everything from `Client` and related wrapper methods.

suluvai replied:
We will not have standard API Requests in the docs anymore.
Will remove this.

suluvai committed:
godoc comments

zatlite commented on imx/interface.go:
```diff
@@ -0,0 +1,107 @@
+	Environment
+}
+
+// Environment holds ethereum network and contract address information. Part of the config.
```
Is this partial sentence a typo?
```suggestion
// Environment holds Ethereum network and contract address information.
```

suluvai committed:
Fixed broken readme sections and updates to signer

suluvai commented:
Whole readme needs a once over clean up and updates. Will address it as a seperate PR.

zatlite commented on README.md:
```diff
@@ -246,40 +224,48 @@ func DepositNft(l1signer immutable.L1Signer, starkKey, assetType, vaultID, token
-    if err != nil {
-        return nil, fmt.Errorf("error in creating L1Signer: %v", err)
-    }
+	l1signer, err := ethereum.NewSigner(signerPrivateKey, cfg.ChainID)
```
indenting all broken again.

zatlite commented on imx/signers/stark/factory.go:
```diff
@@ -11,13 +11,18 @@ import (
 // package-global variable to represent our standard stark curve
-// generally bad practice, but has no state + we never configure it
-// unexported so can be updated in future if necessary
-// TODO: should we do this in init() to allow us to handle the eaten error here
```
I came to see why GenerateKey needs to return error. 
I think the suggestion in this TODO is worth exploring. 
Have you explored it and found a problem therefore removing this TODO comment?

suluvai replied:
Yes, the solution is implemented in the following lines.

suluvai committed:
Dx 1162 (#64)

* Limiting Public API interface (Part 1)

* DX-1162: Added public interface wrapper funtions

* Linter fixes

* Undo unintended change

* Linter fixes and pre-commit hook

* Fxied comments

* id var name update

* Added docs for CreateCollection method

* linter fixes

* DX-1162: Added godoc comments for public methods.

* DX-1162: Updated to use behavioral error pattern.

* DX-1162: golangci-lint install instructions

* DX-1162: Updated to accept undecimalised values for deposit and withdrawal workflows.

* DX-1162: End of file new line

* DX-1162: Added comment for Wei type.

* DX-1162: Removed ToDenomination function as it is not required anymore.

* DX-1162: Removed ToDenomination tests.

* DX-1162: Updated ERC20 tokens to take undecimalised uint64 values.

suluvai committed:
DX-1163: Updated signer to generate random key as per TS SDK. (#68)

* Limiting Public API interface (Part 1)

* DX-1162: Added public interface wrapper funtions

* Linter fixes

* Undo unintended change

* Linter fixes and pre-commit hook

* Fxied comments

* id var name update

* Added docs for CreateCollection method

* linter fixes

* DX-1162: Added godoc comments for public methods.

* DX-1162: Updated to use behavioral error pattern.

* DX-1162: golangci-lint install instructions

* DX-1162: Updated to accept undecimalised values for deposit and withdrawal workflows.

* DX-1162: End of file new line

* DX-1162: Added comment for Wei type.

* DX-1162: Removed ToDenomination function as it is not required anymore.

* DX-1162: Removed ToDenomination tests.

* DX-1162: Updated ERC20 tokens to take undecimalised uint64 values.

* DX-1163: Updated signer to generate random key as per TS SDK.

* DX-1163: Linter fix and removed unused dependencies

* DX-1163: Fixes to grinding logic as per PR feedback

suluvai committed:
Dx 1162 (#70)

* Limiting Public API interface (Part 1)

* DX-1162: Added public interface wrapper funtions

* Linter fixes

* Undo unintended change

* Linter fixes and pre-commit hook

* Fxied comments

* id var name update

* Added docs for CreateCollection method

* linter fixes

* DX-1162: Added godoc comments for public methods.

* DX-1162: Updated to use behavioral error pattern.

* DX-1162: golangci-lint install instructions

* DX-1162: Updated to accept undecimalised values for deposit and withdrawal workflows.

* DX-1162: End of file new line

* DX-1162: Added comment for Wei type.

* DX-1162: Removed ToDenomination function as it is not required anymore.

* DX-1162: Removed ToDenomination tests.

* DX-1162: Updated ERC20 tokens to take undecimalised uint64 values.

* DX-1162: Implemented new methods to create apirequest with mandatory and default values that are not exposed to end user

* DX-1162: Removed IsRegisteredOffChain as discussed to keep the public interface minimal

suluvai committed:
DX-1084: Removed uncessary dependencies. (#69)

suluvai committed:
Dx 1208 (#71)

* DX-1208: WIP commit for examples refactor into their own independent examples

* DX-1208: Fixes for examples on refactoring and moved env file to be independent for each example

* Dx-1208: Minting workflow example tested

* DX-1208: Fixes for examples code

* DX-1208: optimised number of lines of code and addressed core review feedback

suluvai committed:
DX-1242: Updated to use APIError from the openapi spec (#72)

* Linter fixes and pre-commit hook

* Fxied comments

* id var name update

* DX-1162: Updated to accept undecimalised values for deposit and withdrawal workflows.

* DX-1242: Upated to use APIError from the openapi spec and to be in line with TS core sdk

* DX-1242: godoc updated

* DX-1242: File rename from api_error.go to imxerror.go

* DX-1239: Added metedata refresh api endpoints (#73)

* Dx 1243 (#74)

* DX-1239: Added metedata refresh api endpoints

* DX-1243: Exposing contracts including IERC20 and IERC721 contracts

* Dx-1243: PR fedback to function rename

* DX-1261: Updated CompleteEthWithdrawl function to be CompleteWithdraw… (#75)

* DX-1261: Updated CompleteEthWithdrawl function to be CompleteWithdrawal in line with the common agreed interface across SDKs

* DX-1210: Updated readme to reflect the public interface changes (#76)

* DX-1210: Updated readme to reflect the public interface changes

* DX-1210: Updated to use permalink as per code review

suluvai committed:
pr feedback
